homeassistant/components/template/config.py

```diff
@@ -200,9 +198,45 @@ async def _async_resolve_blueprints(
             # CONF_VARIABLES should not be removed because the variables are always
             # executed between the trigger and action.
             if CONF_TRIGGERS not in config and CONF_VARIABLES in config:
-                config[platform][CONF_VARIABLES] = config.pop(CONF_VARIABLES)
+                section_variables = config.pop(CONF_VARIABLES)
+
+                # State based template entities have 2 layers of variables.  Variables at the section level
+                # and variables at the entity level should be merged together at the entity level.
+                config[platform][CONF_VARIABLES] = (
+                    {
+                        **section_variables,
+                        **config[platform].pop(CONF_VARIABLES),
+                    }
+                    if CONF_VARIABLES in config[platform]
+                    else section_variables
+                )
+
         raw_config = dict(config)
 
+    # For Trigger based template entities CONF_VARIABLES should not be altered because the variables
+    # are always executed between the trigger and action.
+    elif CONF_TRIGGERS not in config and CONF_VARIABLES in config:
+        # State based template entities have 2 layers of variables.  Variables at the section level
+        # and variables at the entity level should be merged together at the entity level.
+        section_variables = config.pop(CONF_VARIABLES)
+        platform_config: list[ConfigType] | ConfigType
+        platforms = [platform for platform in PLATFORMS if platform in config]
+        for platform in platforms:
+            platform_config = config[platform]
+            if platform in PLATFORMS:
+                if isinstance(platform_config, dict):
+                    platform_config = [platform_config]
```

Comment:
Why do we need to check if platform is in `PLATFORMS` and handle both list and non-list cases, does this run before config validation?